File: libs/power-log-parser/src/lib/parsers/card-traded-parser.ts
import { BlockType, CardIds, GameTag, Step, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Action, FullEntity, Node, NodeType, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

// TODO: Oracle
const Oracle = {
	FindCardCreator(
		_gameState: GameState,
		_entity: FullEntity,
		_node: Node,
	): [string, number] | null {
		return null;
	},
	PredictCardId(
		_gameState: GameState,
		_creatorCardId: string | null | undefined,
		_creatorEntityId: number,
		_node: Node,
		_cardId: string | null,
	): string | null {
		return null;
	},
};

export class CardTradedParser implements ActionParser {
	readonly ParserName = 'CardTradedParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.TagChange &&
			(node.Object as TagChange).Name === (GameTag.ZONE as number) &&
			(node.Object as TagChange).Value === (Zone.DECK as number) &&
			this.isTrade(node.Parent)
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity)!;
		const zoneInt = entity.GetTag(GameTag.ZONE) === -1 ? 0 : entity.GetTag(GameTag.ZONE);
		const initialZone = Zone[zoneInt] ?? String(zoneInt);
		let cardId: string | null = entity.CardId;
		const controllerId = entity.GetEffectiveController();
		if (!cardId || cardId.length === 0) {
			const creator = Oracle.FindCardCreator(this.GameState, entity, node);
			cardId = Oracle.PredictCardId(this.GameState, creator?.[0], creator?.[1] ?? -1, node, cardId);
		}

		const isBeforeMulligan = this.GameState.GetGameEntity()?.GetTag(GameTag.NEXT_STEP) === -1;
		const isOpponentMulligan =
			this.GameState.GetGameEntity()?.GetTag(GameTag.NEXT_STEP) === (Step.BEGIN_MULLIGAN as number) &&
			entity.GetController() === this.StateFacade.OpponentPlayer?.PlayerId;
		if ((isOpponentMulligan || isBeforeMulligan) && cardId === CardIds.EncumberedPackMule) {
			cardId = '';
		}

		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'TRADE_CARD',
				GameEventHelper.CreateProvider(
					'TRADE_CARD',
					cardId ?? null as any,
					controllerId,
					entity.Id,
					this.StateFacade,
					{
						InitialZone: initialZone,
					},
				),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}

	private isTrade(node: Node | null): boolean {
		return (
			node?.Type === NodeType.Action && (node.Object as Action).Type === (BlockType.DECK_ACTION as number)
		);
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/CardTradedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

namespace HearthstoneReplays.Events.Parsers
{
    public class CardTradedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public CardTradedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.ZONE
                && (node.Object as TagChange).Value == (int)Zone.DECK
                && IsTrade(node.Parent);
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities[tagChange.Entity];
            var zoneInt = entity.GetTag(GameTag.ZONE) == -1 ? 0 : entity.GetTag(GameTag.ZONE);
            var initialZone = ((Zone)zoneInt).ToString();
            var cardId = entity.CardId;
            var controllerId = entity.GetEffectiveController();
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, tagChange, null);
            if (cardId == null || cardId.Length == 0)
            {
                var creator = Oracle.FindCardCreator(GameState, entity, node);
                cardId = Oracle.PredictCardId(GameState, creator?.Item1, creator?.Item2 ?? -1, node, cardId);
            }

            // Because Encumbered Pack Mule reveals itself if drawn during mulligan, we need to 
            // have a special rule
            var isBeforeMulligan = GameState.GetGameEntity().GetTag(GameTag.NEXT_STEP) == -1;
            var isOpponentMulligan = GameState.GetGameEntity().GetTag(GameTag.NEXT_STEP) == (int)Step.BEGIN_MULLIGAN
                && entity.GetController() == StateFacade.OpponentPlayer.PlayerId;
            if ((isOpponentMulligan || isBeforeMulligan) && cardId == CardIds.EncumberedPackMule)
            {
                cardId = "";
            }

            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                "TRADE_CARD",
                GameEvent.CreateProvider(
                    "TRADE_CARD",
                    cardId,
                    controllerId,
                    entity.Id,
                    StateFacade,
                    //gameState,
                    new {
                        InitialZone = initialZone,
                    }),
                true,
                node)
            };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }

        private bool IsTrade(Node node)
        {
            return node?.Type == typeof(Action)
                && (node.Object as Action).Type == (int)BlockType.DECK_ACTION;
        }
    }
}
